src/lib.rs

```diff
@@ -0,0 +1,136 @@
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any person obtaining a
+// copy of this software, associated documentation and/or data (collectively the "Software"), free
+// of charge and under any and all copyright rights in the Software, and any and all patent rights
+// owned or freely licensable by each licensor hereunder covering either (i) the unmodified
+// Software as contributed to or provided by such licensor, or (ii) the Larger Works (as defined
+// below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software is contributed
+// by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create derivative works
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// What was the outcome of a `jit_control_point`?
+#[derive(Debug, PartialEq)]
+pub enum ControlOutcome {
+    /// The `jit_control_point` executed a trace: the interpreter does not need to execute a
+    /// trace for this location in the end user's code.
+    ExecutedTrace,
+    /// The `jit_control_point` did not execute a trace: the interpreter needs to execute the
+    /// relevant code for this location in the end user's code.
+    DidntExecuteTrace
+}
+
+/// The current meta-tracing state of a given location in the end-user's code.
+#[derive(Debug)]
+enum LocCount {
+    /// Below `HOT_THRESHOLD`, but counting up towards it.
+    Counting(u16),
+    /// A trace is being collected.
+    Tracing,
+    /// A compiled trace is available.
+    Trace
+}
+
+/// A meta-tracer.
+pub struct MetaTracer<Loc> {
+    counts: Mutex<HashMap<Loc, LocCount>>,
```

Comment:
Is `Mutex` what we want (it may well be, but I want to ask)? This means only one thread can ever observe the data. Before we had `RwLock` which would allow multiple readers.

Reply:
No, `Mutex` is not a good long-term solution -- but `RwLock` also sucks as it's still a heavyweight solution for something where performance will be critical. But this is a detail for another PR. I have ideas, although they're not completely implemented yet.

Reply:
ok, let's raise issues for these things once we merge this.

Reply:
Give me a chance to finish off the next PR after this one -- or to find that I'm stuck :)

Reply:
[My earlier explanation was a bit terse. I should have said that one problem with `RwLock` is that it doesn't provide an "upgrade" from read to write locks. Indeed, it's impossible to guarantee that any read lock can be upgraded (since two readers could request an upgrade simultaneously). So it's difficult to make a nice API for this. However, the bigger problem is that even the read lock is going to be too expensive for us in the long run, since *mostly* in a long-running program we expect to say "do you have a trace for location L" and the answer to be "yes". i.e. for long-running programs most of the time is going to be spent returning the same answer. So we probably want to use a (lightweight) atomic read rather than a heavy-weight lock. I've sketched a couple of solutions, but have a better one in mind. It'll take a few days.]

Reply:
Understood about upgrading the locks. I had that very same conversation with someone last week!

But note that in the utopian circumstance:

> mostly in a long-running program we expect to say "do you have a trace for location L" and the answer to be "yes"

A `RwLock` won't be terrible, as you will only have readers.

For me the shortcoming of the `RwLock` is where a new trace needs to be built and in order to insert the trace, we would have to wait for N readers to release their read locks.

I'm not saying `RwLock` is the best solution, but I do think it's probably better than a `Mutex` in terms of lock contention for our "expected common case" (however, I always expect the unexpected since the warmup experiment!).

Anyway, I await your new proposals, good luck! :)

Reply:
I'll try and put some numbers on this soon, as my current experiment wasn't exactly rigorous ;)